src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDiffieHellmanAndroid.Derive.cs

```diff
@@ -0,0 +1,196 @@
+                bool disposeOtherKey = false;
+
+                if (otherKey == null)
+                {
+                    disposeOtherKey = true;
+
+                    ECParameters otherParameters =
+                        thisIsNamed
+                            ? otherPartyPublicKey.ExportParameters()
+                            : otherPartyPublicKey.ExportExplicitParameters();
+
+                    otherKey = new ECDiffieHellmanAndroidPublicKey(otherParameters);
+                }
+
+                bool otherIsNamed = otherKey.HasCurveName;
+
+                SafeEcKeyHandle? ourKey = null;
+                SafeEcKeyHandle? theirKey = null;
+                byte[]? rented = null;
+                int secretLength = KeySize;
+
+                try
+                {
+                    if (otherKey.KeySize != KeySize)
+                    {
+                        throw new ArgumentException(SR.Cryptography_ArgECDHKeySizeMismatch, nameof(otherPartyPublicKey));
+                    }
+
+                    if (otherIsNamed == thisIsNamed)
+                    {
+                        ourKey = _key.UpRefKeyHandle();
+                        theirKey = otherKey.DuplicateKeyHandle();
+                    }
+                    else if (otherIsNamed)
+                    {
+                        ourKey = _key.UpRefKeyHandle();
+
+                        using (ECAndroid tmp = new ECAndroid(otherKey.ExportExplicitParameters()))
+                        {
+                            theirKey = tmp.UpRefKeyHandle();
+                        }
+                    }
+                    else
+                    {
+                        using (ECAndroid tmp = new ECAndroid(thisKeyExplicit))
+                        {
+                            ourKey = tmp.UpRefKeyHandle();
+                        }
+
+                        theirKey = otherKey.DuplicateKeyHandle();
+                    }
+
+                    // Indicate that secret can hold stackallocs from nested scopes
+                    Span<byte> secret = stackalloc byte[0];
+
+                    // Arbitrary limit. But it covers secp521r1, which is the biggest common case.
+                    const int StackAllocMax = 66;
+
+                    if (secretLength > StackAllocMax)
+                    {
+                        rented = CryptoPool.Rent(secretLength);
+                        secret = new Span<byte>(rented, 0, secretLength);
+                    }
+                    else
+                    {
+                        secret = stackalloc byte[secretLength];
+                    }
+
+                    if (!Interop.AndroidCrypto.EcdhDeriveKey(ourKey, theirKey, secret, out int usedBufferLength))
+                    {
+                        throw new CryptographicException();
+                    }
+
+                    secret = secret.Slice(0, usedBufferLength);
```

Comment:
Presumably we want to assert that usedBufferLength was the expected value?

If it was shorter than expected then we might have to manually insert some padding zeros... but I'm hoping that the ECDH implementation did that for us.  (I think it's a 50/50 to get a leading 0x00 in secp521r1 ECDH, so shouldn't take long to verify)

Reply:
Should I just run ECDH a few times to see if I get a leading zero with that curve?

Reply:
Yeah.  Make sure that at least one of the two keys varies, though.  Otherwise it'll just produce the same answer over and over again :smile:.

Reply:
Ran this 10 times with randomized keys and never saw a mismatch between `secretLength` and `usedBufferLength`. Added an assert to continue to validate that they match.